### PR TITLE
Site title: Add option to toggle home link

### PR DESCRIPTION
**Pull request by carolinan**
## Description

Partial for https://github.com/WordPress/gutenberg/issues/30918
Previously, the site title was always a link. This adds an option to remove the default link.

## How has this been tested?

Tested manually by adding a site title block:
Text and link color options
With and without link
With and without opening in a new tab.

## Types of changes

Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/site-title/edit/index.js [hunk 1]
```diff
@@ -26,14 +28,49 @@ export default function SiteTitleEdit( {
...
+	const { level, textAlign, isLink, linkTarget } = attributes;
 	const [ title, setTitle ] = useEntityProp( 'root', 'site', 'title' );
 	const TagName = level === 0 ? 'p' : `h${ level }`;
 	const blockProps = useBlockProps( {
 		className: classnames( {
 			[ `has-text-align-${ textAlign }` ]: textAlign,
 		} ),
 	} );
+
+	let titleElement = (
```
Comment (carolinan) [inline packages/block-library/src/site-title/edit/index.js]: How can I reduce this to remove the duplication?

File: packages/block-library/src/site-title/block.json [hunk 2]
```diff
@@ -12,14 +12,21 @@
...
+			"type": "string",
+			"default": "_self"
 		}
 	},
 	"supports": {
 		"align": [ "wide", "full" ],
 		"html": false,
 		"color": {
 			"gradients": true,
-			"text": false,
```
Comment (aristath) [inline packages/block-library/src/site-title/block.json:22]: Why was this line removed?

Comment (carolinan) [reply packages/block-library/src/site-title/block.json:22]: Because when the site title is not a link, the link text color option is not available.
When the site title is plain text, the regular text color option needs to be available, and for that, it can not be set to false.

It does not need to be set to true, because it is enabled by default.

File: packages/block-library/src/site-title/edit/index.js (hunk 1 above)
Comment (aristath) [reply packages/block-library/src/site-title/edit/index.js]: Done in https://github.com/WordPress/gutenberg/pull/31540/commits/1c64ce0cd953b0f095bc92760286ec4488cd8d12 :+1:

**Review by aristath — APPROVED** [review]

**Comment by carolinan** [comment]
Is there a good way to remove the `href=""` from the `div` in the editor?

**Comment by aristath** [comment]
There were a lot of conflicts due to recent tweaks in some blocks (see #32817). As a result, I had to rebase the PR and in the process of resolving conflicts I had to refactor a lot of things.
I tested the PR and it still works with the latest refactor, but I'd be more comfortable if @ntsekouras and @carolinan confirmed we're OK here.

> Is there a good way to remove the `href=""` from the `div` in the editor?

Yes, by returning `undefined` (already done)

**Comment by carolinan** [comment]
The PR refactor is working well for me, the block works as I expect in the editors and front.

There was a merge conflict because the fixtures has moved to test/integration/fixtures/blocks/, so I updated the fixture again to solve it.
But because this block was included in core in 5.8, and this PR changes block attributes, does it not need a deprecation?

**Comment by gziolo** [comment]
As far as I can tell this block is fully dynamic - save content is empty and everything is handled when rendering on the front-end. In that case, you don't have to worry about deprecations.

**Comment by carolinan** [comment]
> As far as I can tell this block is fully dynamic - save content is empty and everything is handled when rendering on the front-end. In that case, you don't have to worry about deprecations.

Thank you, this is confusing because the only documentation I can find for deprecation seems to be for custom blocks. Not for best practices for the plugin/core?

File: packages/block-library/src/site-title/edit/index.js [hunk 3]
```diff
@@ -87,6 +96,28 @@ export default function SiteTitleEdit( {
 					} }
 				/>
 			</BlockControls>
+			<InspectorControls>
+				<PanelBody title={ __( 'Link settings' ) }>
+					<ToggleControl
+						label={ __( 'Make title link to home' ) }
```
Comment (carolinan) [inline packages/block-library/src/site-title/edit/index.js:100]: For the post title the text is: `Make title a link`, it assumes the user knows its a link to the post.
Should this text be the same, or `Make title a link to home` ?
```suggestion
						label={ __( 'Make title a link to home' ) }
```

**Comment by gziolo** [comment]
> > As far as I can tell this block is fully dynamic - save content is empty and everything is handled when rendering on the front-end. In that case, you don't have to worry about deprecations.
> 
> Thank you, this is confusing because the only documentation I can find for deprecation seems to be for custom blocks. Not for best practices for the plugin/core?

The most detailed document is here:

https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-deprecation.md

We should clarify that you need deprecations only when the save implementation outputs something.

**Comment by ntsekouras** [comment]
>As far as I can tell this block is fully dynamic - save content is empty and everything is handled when rendering on the front-end. In that case, you don't have to worry about deprecations.

I'm not sure this is 100% true 🤔 . For example if we renamed an attribute, wouldn't we need a deprecation and handle the `php` part as well? In this case a deprecation is certainly not needed because we add new attributes and with the default value of `isLink`, we match the previous behavior.

File: packages/block-library/src/site-title/edit/index.js (hunk 3 above)
Comment (ntsekouras) [reply packages/block-library/src/site-title/edit/index.js:100]: Could just `Link to home` work?

File: packages/block-library/src/site-title/edit/index.js [hunk 4]
```diff
@@ -63,12 +66,18 @@ export default function SiteTitleEdit( {
...
-			</a>
+			{ isLink ? (
+				<a
+					href="#site-title-pseudo-link"
+					onClick={ ( event ) => event.preventDefault() }
+				>
+					{ readOnlyTitle || __( 'Site Title placeholder' ) }
+				</a>
+			) : (
+				<div style={ { display: 'inline-block' } }>
```
Comment (ntsekouras) [inline packages/block-library/src/site-title/edit/index.js]: Why this `div` is needed? Can we handle this `inline-block` if needed with css instead or at the parent element?

>const TagName = level === 0 ? 'p' : `h${ level }`;

In the case of `level:0` and no link, will add a `div` inside a `p`, which is not valid, no?

File: packages/block-library/src/site-title/edit/index.js [hunk 5]
```diff
@@ -87,6 +96,28 @@ export default function SiteTitleEdit( {
...
 			</BlockControls>
+			<InspectorControls>
+				<PanelBody title={ __( 'Link settings' ) }>
+					<ToggleControl
+						label={ __( 'Make title link to home' ) }
+						onChange={ () => setAttributes( { isLink: ! isLink } ) }
+						checked={ isLink }
+					/>
+					{ isLink && (
+						<>
```
Comment (ntsekouras) [inline packages/block-library/src/site-title/edit/index.js]: Nitpick: Fragment doesn't affect anything but isn't required here..

File: packages/block-library/src/site-title/edit/index.js [hunk 6]
```diff
@@ -49,7 +51,8 @@ export default function SiteTitleEdit( {
 	const siteTitleContent = canUserEdit ? (
 		<TagName { ...blockProps }>
 			<RichText
-				tagName="a"
+				tagName={ isLink ? 'a' : 'div' }
```
Comment (ntsekouras) [inline packages/block-library/src/site-title/edit/index.js]: Same as below about `div` inside a `p`

File: packages/block-library/src/site-title/index.php [hunk 7]
```diff
@@ -25,14 +25,26 @@ function render_block_core_site_title( $attributes ) {
 		$tag_name = 0 === $attributes['level'] ? 'p' : 'h' . $attributes['level'];
 	}
 
-	$link               = sprintf( '<a href="%1$s" rel="home">%2$s</a>', get_bloginfo( 'url' ), $site_title );
+	$title = $site_title;
```
Comment (ntsekouras) [inline packages/block-library/src/site-title/index.php]: Nit: You don't need this extra var.

**Review by ntsekouras — COMMENTED** [review]
Thanks for your work here @carolinan! Besides some small comments, this looks good!

File: packages/block-library/src/site-title/edit/index.js (hunk 4 above)
Comment (carolinan) [reply packages/block-library/src/site-title/edit/index.js]: You are right, `div` is not valid inside a heading or `p`.

**Review by ntsekouras — APPROVED** [review]
This LGTM - thanks!

IMO current `copy` is not confusing and is something that can be even addressed in a follow up. Besides that, everything else seems good.

**Comment by kristastevens** [comment]
Hi @carolinan, 

> carolinan added the Needs Copy Review label 7 hours ago

Thanks for the ping! I think in this case, `Make title link to home` is clear. 

Another slightly more active alternative could be, `Link title to home`

**Comment by carolinan** [comment]
Tests are passing again so I will merge this.
